rpcs3/rpcs3qt/update_manager.cpp

```diff
@@ -0,0 +1,547 @@
+﻿#include "stdafx.h"
+#include "update_manager.h"
+#include "rpcs3_version.h"
+#include "Utilities/StrUtil.h"
+#include "Crypto/sha256.h"
+#include "Emu/System.h"
+
+#include <QMessageBox>
+#include <sstream>
+#include <iomanip>
+
+#if defined(_WIN32)
+#include <windows.h>
+#include <CpuArch.h>
+#include <7z.h>
+#include <7zAlloc.h>
+#include <7zBuf.h>
+#include <7zCrc.h>
+#include <7zFile.h>
+#include <7zVersion.h>
+
+#define PATH_MAX MAX_PATH
+
+#else
+#include <unistd.h>
+#include <sys/stat.h>
+#endif
+
+update_manager::update_manager()
+{
+	m_manager.setRedirectPolicy(QNetworkRequest::NoLessSafeRedirectPolicy);
+}
+
+void update_manager::check_for_updates(bool automatic, QWidget* parent)
+{
+	if (QSslSocket::supportsSsl() == false)
+	{
+		LOG_ERROR(GENERAL, "Can't update! Please make sure your system supports SSL.");
+		if (!automatic)
+			QMessageBox::warning(parent, tr("Auto-updater"), tr("Can't update! Please make sure your system supports SSL."));
+		return;
+	}
+
+	m_parent = parent;
+
+	const std::string request_url = m_update_url + rpcs3::get_commit_and_hash().second;
+	QNetworkReply* reply_json     = m_manager.get(QNetworkRequest(QUrl(QString::fromStdString(request_url))));
+
+	m_progress_dialog = new progress_dialog(tr("Checking For Updates"), tr("Please wait..."), tr("Abort"), 0, 100, parent);
+	m_progress_dialog->setAutoClose(false);
+	m_progress_dialog->setAutoReset(false);
+	m_progress_dialog->show();
+
+	connect(m_progress_dialog, &QProgressDialog::canceled, reply_json, &QNetworkReply::abort);
+	connect(m_progress_dialog, &QProgressDialog::finished, m_progress_dialog, &QProgressDialog::deleteLater);
+
+	// clang-format off
+	connect(reply_json, &QNetworkReply::downloadProgress, [&](qint64 bytesReceived, qint64 bytesTotal)
+	{
+		m_progress_dialog->setMaximum(bytesTotal);
+		m_progress_dialog->setValue(bytesReceived);
+	});
+
+	connect(reply_json, &QNetworkReply::finished, [=]()
+	{
+		handle_reply(reply_json, &update_manager::handle_json, automatic, "Retrieved Json Info");
+	});
+	// clang-format on
+}
+
+bool update_manager::handle_reply(QNetworkReply* reply, std::function<bool(update_manager& man, const QByteArray&, bool)> func, bool automatic, const std::string& message)
+{
+	if (auto error = reply->error(); error != QNetworkReply::NoError)
+	{
+		reply->deleteLater();
+		if (error == QNetworkReply::OperationCanceledError)
```

Comment:
you probably meant != because OperationCanceledError usually means that the user aborted.